ocaml/xapi/message_forwarding.ml

```diff
@@ -3270,15 +3312,22 @@ module Forward = functor(Local: Custom_actions.CUSTOM_ACTIONS) -> struct
       let local_fn = Local.SR.destroy ~sr in
       with_sr_marked ~__context ~sr ~doc:"SR.destroy" ~op:`destroy
         (fun () ->
-           forward_sr_op ~consider_unplugged_pbds:true ~local_fn ~__context ~self:sr
-             (fun session_id rpc -> Client.SR.destroy rpc session_id sr))
+          assert_all_pbds_unplugged ~__context ~sr;
+          assert_sr_not_indestructible ~__context ~sr;
+          assert_sr_not_local_cache ~__context ~sr;
+
+          forward_sr_all_op ~local_fn ~__context ~self:sr
+            (fun session_id rpc -> Client.SR.destroy rpc session_id sr))
 
-    (* don't forward this is just a db call *)
     let forget ~__context ~sr =
       info "SR.forget: SR = '%s'" (sr_uuid ~__context sr);
+      let local_fn = Local.SR.forget ~sr in
       with_sr_marked ~__context ~sr ~doc:"SR.forget" ~op:`forget
         (fun () ->
-           Local.SR.forget ~__context ~sr)
+          assert_all_pbds_unplugged ~__context ~sr;
+
+          forward_sr_all_op ~local_fn ~__context ~self:sr
```

Comment:
What happens if a host is offline/unreachable in the pool? Should we make this function best-effort, or should we have it fail if it cannot reach all hosts?

Reply:
I think that would be best, we don't want the operation to fail while it was successful before.
This will mean that if the forwarding fails for one host, the behaviour of those SR metrics will be the same as before, so it won't introduce a regression, even if it's not ideal.

Customers still have the option to manually delete the metrics if all else fails.